Lombiq.Hosting.Tenants.FeaturesGuard/Handlers/FeaturesEventHandler.cs

```diff
@@ -14,168 +13,137 @@ namespace Lombiq.Hosting.Tenants.FeaturesGuard.Handlers;
-    private readonly IShellDescriptorManager _shellDescriptorManager;
-
-    public FeaturesEventHandler(
-        IShellFeaturesManager shellFeaturesManager,
-        IOptions<ConditionallyEnabledFeaturesOptions> conditionallyEnabledFeaturesOptions,
-        ShellSettings shellSettings,
-        IConfiguration configuration,
-        IShellDescriptorManager shellDescriptorManager)
-    {
-        _shellFeaturesManager = shellFeaturesManager;
-        _conditionallyEnabledFeaturesOptions = conditionallyEnabledFeaturesOptions;
-        _shellSettings = shellSettings;
-        _shellDescriptorManager = shellDescriptorManager;
-    }
-
     Task IFeatureEventHandler.InstallingAsync(IFeatureInfo feature) => Task.CompletedTask;
 
     Task IFeatureEventHandler.InstalledAsync(IFeatureInfo feature) => Task.CompletedTask;
 
     Task IFeatureEventHandler.EnablingAsync(IFeatureInfo feature) => Task.CompletedTask;
 
-    Task IFeatureEventHandler.EnabledAsync(IFeatureInfo feature) => EnableConditionallyEnabledFeaturesAsync(feature);
+    Task IFeatureEventHandler.EnabledAsync(IFeatureInfo feature) => HandleConditionallyEnabledFeaturesAsync();
 
     Task IFeatureEventHandler.DisablingAsync(IFeatureInfo feature) => Task.CompletedTask;
 
-    async Task IFeatureEventHandler.DisabledAsync(IFeatureInfo feature)
-    {
-        await KeepConditionallyEnabledFeaturesEnabledAsync(feature);
-        await DisableConditionallyEnabledFeaturesAsync(feature);
-    }
+    Task IFeatureEventHandler.DisabledAsync(IFeatureInfo feature) => HandleConditionallyEnabledFeaturesAsync();
 
     Task IFeatureEventHandler.UninstallingAsync(IFeatureInfo feature) => Task.CompletedTask; // #spell-check-ignore-line
 
     Task IFeatureEventHandler.UninstalledAsync(IFeatureInfo feature) => Task.CompletedTask;
 
     /// <summary>
-    /// Enables conditional features (key) if one of their corresponding condition features (value) was enabled.
+    /// Enables or disables conditional features depending on ConditionallyEnabledFeaturesOptions.
+    /// Prevents disabling features that should be enabled according to their conditions.
     /// </summary>
-    /// <param name="featureInfo">The feature that was just enabled.</param>
-    public async Task EnableConditionallyEnabledFeaturesAsync(IFeatureInfo featureInfo)
+    private static Task HandleConditionallyEnabledFeaturesAsync()
     {
-        if (_shellSettings.IsDefaultShell() ||
-            _conditionallyEnabledFeaturesOptions.Value.EnableFeatureIfOtherFeatureIsEnabled is not { } conditionallyEnabledFeatures)
-        {
-            return;
-        }
-
-        var allConditionFeatureIds = new List<string>();
-        foreach (var conditionFeatureIds in conditionallyEnabledFeatures.Values)
-        {
-            allConditionFeatureIds.AddRange(conditionFeatureIds);
-        }
-
-        if (!allConditionFeatureIds.Contains(featureInfo.Id))
+        ShellScope.AddDeferredTask(async scope =>
```

Comment:
Just before, try to use and check a `_DeferredTask` boolean as suggested.

It prevents many deferred tasks from being triggered.

Reply:
@jtkech I tried the `_deferredTaskExecuted` boolean like you have suggested but there is a catch for me with it.
So it basically prevents the handler to run multiple times that is understandable but this for me kills the functionality of the half feature.
For example `OrchardCore.Facebook` and `OrchardCore.Twitter` will be enabled if `Lombiq.UIKit` is enabled.
But after this point if I disable `Lombiq.UIKit` the other 2 features will remain enabled. Also If I want to turn off Facebook I can turn it off eventhough it should stay enabled because `Lombiq.UIKit` is still enabled.
But the handler is not running once more because of the boolean set to true.

Reply:
This delegate handles any number of features to be enabled/disabled, no? So it shouldn't require running more than once for an updated shell.

Reply:
@Piedone The problem if it runs once than as I described 

> Yes, using a local bool (to reduce the deferred tasks) works well on my side (if the handler is a scoped service), if @Psichorex has an use case that doesn't work we would need to understand what happens.
> 
> ```
> if (_deferredTask)
> {
>     return Task.CompletedTask;
> }
> 
> _deferredTask = true;
> 
> ShellScope.AddDeferredTask(async scope =>
> ```

@jtkech Here is my repro if I use `_deferredTask` boolean:
1) Create a new tenant from the default tenant
2) Setup the new tenant
3) When clicking setup the Handler will run and it will enable all conditional features and `_deferredTask` will be true.
4) Go to the new Tenant admin and try to disable a condition feature or a conditional feature
5) The handler for me never runs again thus I can disable a conditional feature eventhough it's condition is still on and it should be kept enabled.

Reply:
Thanks for the repro, I will try it this night.

Can you show me the code that registers the handler and defines the `_deferredTask` variable.

Reply:
@jtkech Sending you what you need.

![image](https://github.com/Lombiq/Hosting-Tenants/assets/92299130/6b128cd8-e463-4030-8bbb-a093ecd90d34)
![image](https://github.com/Lombiq/Hosting-Tenants/assets/92299130/a7e7efa7-9e7c-4c48-a887-7c8b1d79efd0)
![image](https://github.com/Lombiq/Hosting-Tenants/assets/92299130/0313fda0-88f8-4bfb-bddf-d569ad04a31a)

Reply:
For me this is preventing the handler to ever be executed again.
So basically this will execute only once when the tenant is being initialized.

Reply:
Okay just retried, I can't repro, it works well on my side, `_deferredTask` is false again when the handler is resolved from another scope. Did you retry it with the new code where states are retrieved and all the logic done in the deferred task?

I used the same code unless that I always configure the options by code (one conditional/condition) and that I always register the handler by code, looks like you are doing it through a module/feature startup.

Okay, just retried by using a separate `OC.FeaturesGuard` that I included in the setup recipe, all still works fine, the only difference is that I still always config my testing options by code (not from the configuration) in the module startup.

As a side note, not tried but I think you could get rid of the `ConditionallyEnabledFeatures` config section and then bind directly on `ConditionallyEnabledFeaturesOptions`, but as you want.